src/libmw/src/mmr/Segment.cpp

```diff
@@ -0,0 +1,147 @@
+#include <mw/mmr/Segment.h>
+#include <mw/mmr/LeafSet.h>
+#include <mw/mmr/MMR.h>
+#include <mw/mmr/MMRUtil.h>
+
+using namespace mmr;
+
+Segment SegmentFactory::Assemble(const IMMR& mmr, const ILeafSet& leafset, const LeafIndex& first_leaf_idx, const uint16_t num_leaves)
+{
+    if (!leafset.Contains(first_leaf_idx)) {
+        return {};
+    }
+
+    Segment segment;
+    segment.leaves.reserve(num_leaves);
+
+    mmr::LeafIndex leaf_idx = first_leaf_idx;
+    mmr::LeafIndex last_leaf_idx = first_leaf_idx;
+    while (segment.leaves.size() < num_leaves && leaf_idx < leafset.GetNextLeafIdx()) {
+        if (leafset.Contains(leaf_idx)) {
+            last_leaf_idx = leaf_idx;
+            segment.leaves.push_back(mmr.GetHash(leaf_idx.GetNodeIndex()));
+        }
+
+        ++leaf_idx;
+    }
+    
+    const uint64_t num_nodes = leafset.GetNextLeafIdx().GetPosition();
+    std::vector<Index> peak_indices = MMRUtil::CalcPeakIndices(num_nodes);
+    assert(!peak_indices.empty());
+
+    // Populate hashes
+    std::set<Index> hash_indices = CalcHashIndices(leafset, peak_indices, first_leaf_idx, last_leaf_idx);
+    segment.hashes.reserve(hash_indices.size());
+    for (const Index& idx : hash_indices) {
+        segment.hashes.push_back(mmr.GetHash(idx));
+    }
+
+    // Determine the lowest peak that can be calculated using the hashes we've already provided
+    auto peak = *std::find_if(
+        peak_indices.begin(), peak_indices.end(),
+        [&last_leaf_idx](const Index& peak_idx) { return peak_idx > last_leaf_idx.GetNodeIndex(); });
+
+    // Bag the next lower peak (if there is one), so the root can still be calculated
+    segment.lower_peak = BagNextLowerPeak(mmr, peak_indices, peak, num_nodes);
+
+    return segment;
+}
+
+std::set<Index> SegmentFactory::CalcHashIndices(
+    const ILeafSet& leafset,
+    const std::vector<Index>& peak_indices,
+    const mmr::LeafIndex& first_leaf_idx,
+    const mmr::LeafIndex& last_leaf_idx)
+{
+    std::set<Index> proof_indices;
+
+    // 1. Add peaks of mountains to the left of first index
+    boost::optional<Index> prev_peak = boost::make_optional(false, Index());
+    for (const Index& peak_idx : peak_indices) {
+        if (peak_idx < first_leaf_idx.GetNodeIndex()) {
+            proof_indices.insert(peak_idx);
+            prev_peak = peak_idx;
+        } else {
+            break;
+        }
+    }
+
+    // 2. Add indices needed to reach left edge of mountain
+    auto on_mountain_left_edge = [prev_peak](const Index& idx) -> bool {
+        const uint64_t adjustment = !!prev_peak ? prev_peak->GetPosition() + 1 : 0;
+        return ((idx.GetPosition() + 2) - adjustment) == (uint64_t(2) << idx.GetHeight());
+    };
+
+    Index idx = first_leaf_idx.GetNodeIndex();
+    while (!on_mountain_left_edge(idx)) {
+        Index sibling_idx = idx.GetSibling();
+        if (sibling_idx < idx) {
+            proof_indices.insert(sibling_idx);
```

Comment:
What if sibling_idx is a pruned leaf, does that still work?

Reply:
Once *both* siblings are spent (if the siblings are leaves) or marked as pruned parents (in the case of parent indices), those siblings can then be compacted (removed), and the parent of those siblings then gets marked as a pruned parent. So siblings are always compacted together.

P.S. I need to more precisely define the PMMR terminology somewhere. I'll make that a priority.